### ISSUE TITLE
plugins.tml doesn't work with split up packages

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10604]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#906|
|Type|defect|
|Created|2005-02-27 13:48:05Z|

<details><summary>Searchable metadata</summary>

```
trac-id__906 906
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1109512085000000 1109512085000000
changetime__1110730599000000 1110730599000000
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__spiv cc__itamarst cc__Tv cc__jknight cc__zooko
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
getPlugIns only traverses top-level package directories and only looks for one
file in each.  This conflicts strongly with distributing different parts of a
single top-level package separately.

getPlugIns needs to change.  Here's one possible scheme:

  /etc/twisted/plugins.d/ contains files, each containing one or more register()
(or the like) calls.  Each package can put a new file into this directory. 
Additionally, ~/.twisted/plugins.d/ could be searched.  TWISTEDPLUGINPATH could
be examined to find additional locations.  On Windows, the registry could be
used (or something, who gives a crap about Windows anyway?).

  The getPlugIns should remain unaffected by this change, but maybe it would be
desirable to deprecate it in favor of a new one anyway, to reduce confusion in
the switch-over.

  As a bonus, searching 1 or 2 directories (in most cases) should be vastly
faster than the current scheme which has to look for 6 files in at least 8
directories (with an average probably closet to 15).  That's 2 listdirs()
instead of 90 fstat()s.

  This should be done before 2.0.

  Thoughts?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)|@zooko commented|
|-|-|

```
#!html
<pre>
I think it is a mistake that this is assigned to me.  I hardly know 
what you are talking about!

Regards,

Zooko
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Tv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Tv commented|
|-|-|

```
#!html
<pre>
I quite like exarkun's suggestion. But do note that

&lt;hat organization="debian">
Files in /etc are configuration files. That means the admin expects to be able
to modify them without suffering. Files in /usr are not configuration files, and
admins should never modify them. Also, files that clearly are not configuration
files should not be in /etc.

The suggested change has implications: it makes .tml files configuration files.
However, a trivial change to /usr/share/twisted/plugins.d/ will avoid that, if
wanted.
&lt;/hat>

Also, I needed to have .tml files in subpackages, so I wrote this:

import MyTopLevelPackageNameHere
def findTMLs():
    for top in MyTopLevelPackageNameHere.__path__:
        for dirpath, dirnames, filenames in os.walk(top):
            # filter out files we obviously don't need to check - ones with '.'
in them
            dirnames[:] = [s for s in dirnames if "." not in s]

            for filename in filenames:
                if (filename.endswith('.tml')
                    and not filename.startswith('.')
                    and not filename.startswith('#')
                    and filename != 'plugins.tml'):
                    yield os.path.join(dirpath, filename)

def getMorePlugIns(pluginType):
    tmlFiles = list(findTMLs())
    if not tmlFiles:
        return []
    return twplug.loadPlugins(pluginType, tmlFiles)
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
exarkun and I have implemented a new plugin system in Mantissa (the Divmod
appserver) which I think can be rolled back into Twisted pretty shortly.  It's here:
http://www.divmod.org/cvs/branches/glyph/nav-2684/mantissa/plugin.py?rev=7905&amp;view=markup

Basically it abandons the idea of special file types and a special directory,
favoring instead integration with sys.path and import.  It also uses interfaces
rather than string-named plugin types.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
I _don't_ want this change to happen for 2.0, which I plan to release very 
shortly. Our plugins system does function sufficiently for the release to 
go out, and this is way too major a change to make right before the 
release.

I do agree with the general sentiment that our plugin system sucks, though, 
and the solutions outlined seem pretty good, just not before 2.0.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
It sounds like radix meant to downgrade this.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
As long as there are regular releases after 2.0, I don't much care: however,
it's worth noting that the new plugin system can be used side-by-side with
plugins.tml.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
From the original comment on this issue:

"""
getPlugIns only traverses top-level package directories and only looks for one
file in each.  This _conflicts strongly_ with distributing different parts of a
single top-level package separately.
""" 

(emphasis added)

What alternate mechanism has been decided upon to allow all of the Twisted
subpackages which will be created by the 2.0 release to install plugins (for
example, tap plugins)?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Alternate mechanism: keeping the registrations in the toplevel plugins.tml, 
as they are right now.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
What foom said. The set of subprojects is not volatile, and the tap plugins 
function in my 2.0 prereleases.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
This works as expected when certain subpackages are not installed?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="REFERENTIAL INTEGRITY FAILURE's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|REFERENTIAL INTEGRITY FAILURE commented|
|-|-|

```
#!html
<pre>
The issue in that case is that it still shows the subpackage in the
list in "mktap --help", and if you try to mktap something that isn't
there, it'll raise a traceback.

On Thu, 10 Mar 2005 01:56:25 +0000, Glyph Lefkowitz [Twisted issue
tracker] &lt;twisted.roundup@...> wrote:
> 
> Glyph Lefkowitz &lt;glyph@...> added the comment:
> 
> This works as expected when certain subpackages are not installed?
> 
> _________________________________________________________
> Twisted issue tracker &lt;twisted.roundup@...>
> &lt;http://twistedmatrix.com/bugs/#906>
> _________________________________________________________
> 
>
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Downgrading to bug again. Agree with radix it is too late for a major 
change like this for 2.0. It should be revisited for 2.1.

Behavior of mktap is unfortunate but not critical.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Woo trampoline.

Discussed this with Glyph and Itamar.  We all think this should go in for 2.0. 
I will work on it this weekend in time for it to be included in the next
pre-release.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
please explain
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
oops, that was exarkun who said that, not jknight. I take it you're still 
going on the same arguments?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
Resolved by r13168
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
